File: 2841-maximum-sum-of-almost-unique-subarray/2841-maximum-sum-of-almost-unique-subarray.ts
function maxSum(nums: number[], m: number, k: number): number {
  let left = 0;
  let sum = 0;
     let freq = new Map();
    let maxSum = 0;
    for(let j=0;j<nums.length;j++){
     if (!freq.has(nums[j])) {
      freq.set(nums[j], 0);
    }
    freq.set(nums[j], freq.get(nums[j]) + 1);
        sum+=nums[j];
        while(j-left+1>k){
            freq.set(nums[left], freq.get(nums[left]) - 1);
            if(freq.get(nums[left]) === 0){
                freq.delete(nums[left]);
            }
            sum-=nums[left];
            left++;
        }
  let size = freq.size;
        if(size >= m){
            maxSum = Math.max(maxSum,sum);
        }
    }
    return maxSum;
};
// only choose the k elements and keep sliding 

File: 2841-maximum-sum-of-almost-unique-subarray/2841-maximum-sum-of-almost-unique-subarray.cpp
class Solution {
public:
    long long maxSum(vector<int>& nums, int m, int k) {
        long long n = size(nums);
        long long currentSum = 0;
        long long maxSum = 0;
        unordered_map<int,int> freq;
        long long left=0;
        for(int j=0;j<nums.size();j++){
            freq[nums[j]]++;
            currentSum+=nums[j];
            while(j-left+1>k){
                freq[nums[left]]--;
                if(freq[nums[left]] == 0){
                    freq.erase(nums[left]);
                }
                currentSum -= nums[left];
                left++;
            }
            if(freq.size()>=m){
                maxSum = max(maxSum,currentSum);
            }
        }
        return maxSum;
    }
};